File: bench/py/64_struct_create.py
# struct creation: construct structs STANDALONE (not into an array) in a loop,
# read their fields, and accumulate - measures the construction + field-read
# path. Distinct from 58_structs, which builds a list via append; here each
# Point/Vec3 is a fresh standalone value. __slots__ makes CPython instances
# compact, the fairest comparison to MyLang's POD structs.
import sys

scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1


class Point:
    __slots__ = ("x", "y")

    def __init__(self, x, y):
        self.x = x
        self.y = y


class Vec3:
    __slots__ = ("x", "y", "z")

    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z


N = 500000 * scale
sx = 0
fs = 0.0

for i in range(N):
    p = Point(i, i * 2)                       # standalone int-POD construction
    sx += p.x + p.y
    v = Vec3(i * 1.0, i * 2.0, i * 3.0)       # standalone float-POD construction
    fs += v.x + v.y + v.z

print("result:", sx, int(fs))

File: bench/cpp/64_struct_create.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/64_struct_create.my: construct POD structs
 * STANDALONE (not into an array) in a loop, read their fields, accumulate. An
 * int-POD Point and a float-POD Vec3, one of each per iteration. -O3 inlines
 * the construction+field-read to arithmetic - that IS the C++ ceiling. The
 * float accumulation fs is non-associative, so the compiler keeps the loop
 * sequential (O(N)); bench_sink(sx) blocks close-forming the int sum. */
#include "bench.h"

struct Point {
    long x;
    long y;
};

struct Vec3 {
    double x;
    double y;
    double z;
};

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);
    long N = 500000L * scale;
    long sx = 0;
    double fs = 0.0;

    for (long i = 0; i < N; i++) {
        Point p{i, i * 2};                          /* int-POD construction */
        /* BENCH-FAIR (class B): without the escape, SRoA dissolved both
         * structs into registers (asm-verified: no stores) - MyLang
         * materializes a real object (H1 reuses its bytes, but the object
         * and its field stores exist). Escaping the address forces the
         * construction to memory, the same logical shape. */
        asm volatile("" : : "r"(&p) : "memory");
        sx += p.x + p.y;
        bench_sink(sx);
        Vec3 v{i * 1.0, i * 2.0, i * 3.0};          /* float-POD construction */
        asm volatile("" : : "r"(&v) : "memory");
        fs += v.x + v.y + v.z;
    }

    printf("result: %ld %ld\n", sx, (long)fs);
    return 0;
}
